### DISCUSSION TITLE
constructor base call without empty curly brackets

### DISCUSSION THREAD

**Discussion by marcosikkens** [dotnet/csharplang#4309] [Language Ideas]
When writing C# constructors with a base call the following code is often created:

```
public class TestB : TestA
{
 
      public TestB(TestA a): base(a)
      {
          //nothing happens here
      }
 
}

public class TestA
{
   public TestA(TestA a)
   {
       //do something with the members ie:
       this.Something = a.Something;
   }
} 
```

The main thing about this is that sometimes the curly brackets in TestB do not contain any code. it would be nice to be able to just remove those. Like in this example:

```
public class TestB : TestA
{
      public TestB(TestA a): base(a)
 }
```

An alternative I currently see used a lot in this case is the following but it just adds a bit of clutter to the end of the line in my opinion:

```
public class TestB : TestA
{
 
      public TestB(TestA a): base(a)   {  }
 
}
```
If you have any questions (or if it isn't clear) please do not hesitate to ask.

**Comment by Phantonia** [comment]
To help the parser, this would only be done by completing the constructor declaration with a semicolon, like so:
```cs
public TestB(TestA a) : base(a);
```
And does this really save much over
```cs
public TestB(TestA a) : base(a) { }
```

Additionally, ending a method declaration in a semicolon has a clear meaning in C#. It is done for abstract, partial or extern methods, all of which do have a body, it is just not specified at the point where the method is declared.
Using a semicolon to end an empty "method" is not really clear, because it implies a body coming from somewhere else, whereas actually it has a body, it is just empty. The curly brackets clearly communicate that the body is empty.

**Reply by marcosikkens** [reply]
Sorry for the late reply, but I've been thinking about your answer.

Personally I like the semicolon at the end of your first example. It looks a bit cleaner in my honest opinion.

> Additionally, ending a method declaration in a semicolon has a clear meaning in C#. It is done for abstract, partial or extern methods, all of which do have a body, it is just not specified at the point where the method is declared.

In this case the body is actually declared somewhere else (in the base class) so I see no problem in ending with a semicolon in this example. Just to be clear, I want the curly braces when you actually need them but only in that case.

**Reply by KieranDevvs** [reply]
> Using a semicolon to end an empty "method" is not really clear, because it implies a body coming from somewhere else, whereas actually it has a body, it is just empty. The curly brackets clearly communicate that the body is empty.

I think its perfectly clear. Where else would the body be coming from?

Take away inheritance for a second, if you don't have a constructor, is that unclear? Because it has a constructor, its just empty... Its also not like people can some how screw up their code if they don't understand this feature, as the body is optional. People who are used to creating a blank body can still do so.

**Reply by CyrusNajmabadi** [reply]
> I think its perfectly clear. 

I disagree.

> if you don't have a constructor, is that unclear? 

Yes :)

> People who are used to creating a blank body can still do so.

It literally saves one character.  That's it.  You can just write `{}` *today*.  It's been supported since C#1.0.  It's trivially understood as 'has an empty body'.  Why do we need a 1-char version of that?

**Reply by KieranDevvs** [reply]
> > if you don't have a constructor, is that unclear?
> 
> Yes :)

You've just contradicted your own point...

**Reply by CyrusNajmabadi** [reply]
I thought you wrote constructor body.

**Reply by marcosikkens** [reply]
My goal never was to write less characters. I think the expression without the braces is just clearer and causes less clutter. 

We can do a similar thing with automatic usings and namespaces now. So it can be seen as a extension to that.

**Comment by colejohnson66** [comment]
While one could argue that records and primary constructors give precedence to empty method bodies (from their ability to have empty type definitions), I, personally, do not like the idea of empty methods being denoted with a single semicolon. It could also conflict with default interface methods (DIMs) where a `;` denotes an abstract method but `{}` denotes a default implementation that does nothing.

**Reply by marcosikkens** [reply]
I see your point however, you also do not use {} in primary constructors to indicate an empty body. This way you can also have multiple constructors in a class which you cannot do with a primary constructor.

**Comment by nogic1008** [comment]
This proposal is useful for classes that use a Primary Constructor on C#12.

```csharp
public class Sample(int id, string name)
{
    public Sample(): this(0, "");
}
```

The difference between `;` and `{}` may seem like only one character, but there is a big difference in lint and format.

```csharp
    // Editor might be written:
    public Sample(): this(0, "") {}

    // With whitespace...?
    public Sample(): this(0, "") { }

    // Or line break...?
    public Sample(): this(0, "")
    {
    }
```

**Reply by CyrusNajmabadi** [reply]
> but there is a big difference in lint and format.

These are not the concern of the language. :-)

**Reply by HaloFour** [reply]
> but there is a big difference in lint and format.

Those would be concerns to take up with that tooling.  The language isn't going to create new syntactic forms to combat overzealous formatters that try to insert newlines into that construct.

**Reply by nogic1008** [reply]
I don't talk about lint tools, but developer's choices. ("Editor" means not code editor, sorry my poor English...)

This proposal also helps to delegate the main processing to another constructor.
Consider refactoring the following class:

```csharp
public class Sample
{
    private int _x;
    private int _y;

    public Sample()
    {
        (_x, _y) = (0, 0);
        ShouldBeCalled();
    }

    public Sample(int x)
    {
        (_x, _y) = (x, 0);
        ShouldBeCalled();
    }

    public Sample(int x, int y)
    {
        (_x, _y) = (x, y);
        ShouldBeCalled();
    }
}
```

```csharp
public class Sample
{
    private int _x;
    private int _y;

    // Oops, missing `:this(0, 0)`!
    public Sample() { } // No compile error, but ShouldBeCalled() doesn't call
    public Sample(); // This causes compile error because "no body"

    public Sample(int x): this(x, 0) { }
    public Sample(int x): this(x, 0); // Proposal

    public Sample(int x, int y)
    {
        (_x, _y) = (x, y);
        ShouldBeCalled();
    }
}

**Reply by HaloFour** [reply]
> I don't talk about lint tools, but developer's choices.

That would be a stylistic choice on the part of the developer.

> This proposal also helps to delegate the main processing to another constructor.

I don't see how this proposal does that.  Allowing a semicolon only in the case of a chained constructor call would be weirdly inconsistent.

**Reply by nogic1008** [reply]
So, how about the extension that lets you use `this` or `base` keyword in expression body?

```csharp
public Sample() => this(0, 0);
```

This more clarify that the main process is in another constructor.

**Reply by KieranDevvs** [reply]
> Allowing a semicolon only in the case of a chained constructor call would be weirdly inconsistent.

Id argue that its never been more relevant.
C# 8's "simple" using statements, a stylisation change by allowing the replacement of a scope in favour of adding a semi colon.
C# 10's file scoped namespaces, a stylisation change by allowing the replacement of a scope in favour of adding a semi colon.

**Reply by CyrusNajmabadi** [reply]
These feet like entirely different situations. In other words, the other cases are all about nested children not needing delimiters. But that's not the case with constructors.

**Reply by KieranDevvs** [reply]
> These feet like entirely different situations. In other words, the other cases are all about nested children not needing delimiters. But that's not the case with constructors.

Ok? The features all do different things and for different reasons but I think you missed the point that both of these features were added with one of the reasons being:
**Omitting semantics because you don't need them.**

If we look at the language docs under the motivation header for both of these features, here's what we find:
- Simple using statements
  - `"This syntax burden is enough that most coding style guidelines explicitly have an exception around braces for this scenario."`
  - `"The using declaration removes much of the ceremony here and gets C# on par with other languages that include resource management blocks."`
- File scoped namespaces,
  - `"However, both these forms force the user to indent the majority of their code and add a fair amount of ceremony for what is effectively a very basic concept."`
  - `"The primary goal of the feature therefore is to meet the needs of the majority of the ecosystem with less unnecessary boilerplate."`

So I don't understand how in this case of omitting semantics when its not needed, is some how "confusing" yet we do it in other places in the language and no one seems to have an issue.

**Reply by CyrusNajmabadi** [reply]
> but I think you missed the point that both of these features were added

I was the direct author of one of those features, and Co designed the other :-)

I know why they were added. I'm saying that the analogy doesn't hold up here as they aimed to allow avoiding delimiters around nested children. But that's not the benefit here. Here, you only get the benefit when you *only" have `{}`. Making this a savings of just a single character in a very specific case.

**Reply by KieranDevvs** [reply]
> > but I think you missed the point that both of these features were added
> 
> I was the direct author of one of those features, and Co designed the other :-)
> 
> I know why they were added. I'm saying that the analogy doesn't hold up here as they aimed to allow avoiding delimiters around nested children. But that's not the benefit here. Here, you only get the benefit when you *only" have `{}`. Making this a savings of just a single character in a very specific case.

I buy the argument that it only saves one character and might not be worth the effort of adding to the language, that's not what I'm disputing. my issue was with the statement:
>Allowing a semicolon only in the case of a chained constructor call would be weirdly inconsistent.

Or the fact that its somehow confusing.

Its definitely not inconsistent or confusing in my opinion.

**Reply by marcosikkens** [reply]
Part of the motivation as quoted by KyranDavies is to avoid boilerplate. You are saying they are to avoid nested delimiters which is fine. However not part of the motivation.

I am not a language expert and merely a user. But why was the 'lambda' variant added for the getters and the setters. That also saved 2 characters. So why there and not here? I mean if a property can be simplified why not a constructor.  

I assume that nested in your remark means in a method. This is also not in a method.

**Reply by CyrusNajmabadi** [reply]
> That also saved 2 characters

It saves more than that:

```c#
get => 0; // vs
get { return 0; }
```

**Reply by HaloFour** [reply]
> That also saved 2 characters.

And a `return` keyword in the case of getters.

Expression-bodied members work consistently across different member types, including constructors, with or without chained constructors.  If empty statement bodies were permitted, I would expect the same.

**Reply by CyrusNajmabadi** [reply]
>  I mean if a property can be simplified why not a constructor.

We care if we feel the use case is enormously common, the savings significant enough, and the new syntax substantively desirable.  

We do this on a case by case basis. Properties are in huge use. These patterns are super super common with them. And the resultant syntax something we find very attractive.  As such, it read worthwhile. But it's not carte Blanche for every simplification.

**Reply by CyrusNajmabadi** [reply]
> I assume that nested in your remark means in a method.

It does not. It just means "children" in a syntactic sense.  The two features mentioned were about benefitting use cases where the constructs *do* have syntactic children.  That's not the case for *this* feature. This feature only benefits you when you do *not* have syntactic children.

**Reply by CyrusNajmabadi** [reply]
> Its definitely not inconsistent or confusing in my opinion.

It's confusing to me. There are many cases where we do not require the constructor initializer. But now we would, just to be able to use a semicolon for the constructor body.  This is very different from all the other places we allow a semicolon body for other constructs.

**Reply by CyrusNajmabadi** [reply]
> If empty statement bodies were permitted, I would expect the same

Agreed.  I would be more in favor of this if it was used consistently (like how we made it so that `=>` can be used consistently).  Having it *only* for constructors, and *only* when they have an initializer makes it both too niche, and too inconsistent for me to want it.

**Reply by KieranDevvs** [reply]
> > Its definitely not inconsistent or confusing in my opinion.
> 
> It's confusing to me. There are many cases where we do not require the constructor initializer. But now we would, just to be able to use a semicolon for the constructor body. This is very different from all the other places we allow a semicolon body for other constructs.

I'm not sure I understand your example, why would this feature result in constructor initializers being required in some cases?
To my understanding of what is described by the original poster, this is purely syntactic sugar and there would be zero behavioural changes.

**Reply by CyrusNajmabadi** [reply]
The request above said:

> `public Sample(); // This causes compile error because "no body"`

They want semicolon here so it can actually error if they leave off the initializer.  Something they don't get with a normal block body.

**Reply by CyrusNajmabadi** [reply]
> what is described by the original poster, this is purely syntactic sugar

But it is inconsistently applied.  If it represents an empty body, why is it just for constructors?

**Reply by marcosikkens** [reply]
Thanks for the explanation, so let's just say. We remove the semicolon and allow it for other methods. 

What would a good example be if allowing this for other methods? The best thing I can think about are constructors.

**Reply by CyrusNajmabadi** [reply]
Any method that ends up having an empty body because it doesn't need to do anything
